src/imitation/algorithms/bc.py

```diff
@@ -357,48 +397,57 @@ def train(
                 even if `.train()` logged to Tensorboard previously. Has no practical
                 effect if `.train()` is being called for the first time.
         """
-        it = EpochOrBatchIteratorWithProgress(
-            self._demo_data_loader,
-            n_epochs=n_epochs,
-            n_batches=n_batches,
-            on_epoch_end=on_epoch_end,
-            on_batch_end=on_batch_end,
-            progress_bar_visible=progress_bar,
+        if reset_tensorboard:
+            self._bc_logger.reset_tensorboard_steps()
+
+        compute_rollout_stats = RolloutStatsComputer(
+            log_rollouts_venv,
+            log_rollouts_n_episodes,
         )
 
-        if reset_tensorboard:
-            self.tensorboard_step = 0
+        def _on_epoch_end(epoch_number: int):
+            if isinstance(batches_with_stats, tqdm.tqdm):
+                total_num_epochs_str = f"of {n_epochs}" if n_epochs is not None else ""
+                batches_with_progress_bar: tqdm.tqdm = batches_with_stats
+                batches_with_progress_bar.display(
+                    f"Epoch {epoch_number} {total_num_epochs_str}",
+                    pos=1,
+                )
+            self._bc_logger.log_epoch(epoch_number)
```

Comment:
So the key `bc/epoch` only gets logged at the end of each epoch? Seems more intuitive to me to log it at each iteration (with epoch number only incrementing periodically).

Reply:
True, there were even other issues with epoch logging. I fixed it.